core/lib/zksync_core/src/state_keeper/io/seal_logic/mod.rs

```diff
@@ -309,154 +319,74 @@ impl L2BlockSealCommand {
     /// one for sending fees to the operator).
     ///
     /// `l2_shared_bridge_addr` is required to extract the information on newly added tokens.
-    async fn seal_inner(
-        &self,
-        storage: &mut Connection<'_, Core>,
-        is_fictive: bool,
-    ) -> anyhow::Result<()> {
-        self.ensure_valid_l2_block(is_fictive)
-            .context("L2 block is invalid")?;
-
-        let mut transaction = storage.start_transaction().await?;
-        if self.pre_insert_txs {
-            let progress = L2_BLOCK_METRICS.start(L2BlockSealStage::PreInsertTxs, is_fictive);
-            self.insert_transactions(&mut transaction)
-                .await
-                .context("failed persisting transactions in L2 block")?;
-            progress.observe(Some(self.l2_block.executed_transactions.len()));
-        }
-
-        let l1_batch_number = self.l1_batch_number;
-        let l2_block_number = self.l2_block.number;
+    async fn seal_inner(self, pool: ConnectionPool<Core>, is_fictive: bool) -> anyhow::Result<()> {
         let started_at = Instant::now();
-        let progress = L2_BLOCK_METRICS.start(L2BlockSealStage::InsertL2BlockHeader, is_fictive);
+        let this = Arc::new(self);
+        this.ensure_valid_l2_block(is_fictive)
+            .context("L2 block is invalid")?;
 
-        let (l1_tx_count, l2_tx_count) = l1_l2_tx_count(&self.l2_block.executed_transactions);
+        let (l1_tx_count, l2_tx_count) = l1_l2_tx_count(&this.l2_block.executed_transactions);
         let (writes_count, reads_count) =
-            storage_log_query_write_read_counts(&self.l2_block.storage_logs);
+            storage_log_query_write_read_counts(&this.l2_block.storage_logs);
         tracing::info!(
             "Sealing L2 block {l2_block_number} with timestamp {ts} (L1 batch {l1_batch_number}) \
              with {total_tx_count} ({l2_tx_count} L2 + {l1_tx_count} L1) txs, {event_count} events, \
              {reads_count} reads, {writes_count} writes",
-            ts = display_timestamp(self.l2_block.timestamp),
+            l2_block_number = this.l2_block.number,
+            l1_batch_number = this.l1_batch_number,
+            ts = display_timestamp(this.l2_block.timestamp),
             total_tx_count = l1_tx_count + l2_tx_count,
-            event_count = self.l2_block.events.len()
+            event_count = this.l2_block.events.len()
         );
 
-        let definite_vm_version = self
+        if this.pre_insert_txs {
+            let mut connection = pool.connection_tagged("state_keeper").await?;
+            let progress = L2_BLOCK_METRICS.start(L2BlockSealStage::PreInsertTxs, is_fictive);
+            this.insert_transactions(&mut connection)
```

Comment:
I'm going to merge `pre_insert_txs` and `mark_txs_as_executed_in_l2_block` in a separate PR.